sycl/doc/extensions/experimental/sycl_ext_intel_fpga_task_sequence.asciidoc

```diff
@@ -0,0 +1,477 @@
+if the application mistakenly tries to use it on an unsupported backend.
+When this is the case, include a paragraph like this:_
+
+The APIs in this extension may be used only on a device that has
+`aspect::ext_oneapi_foo`.  The application must check that the device has
+this aspect before submitting a kernel using any of the APIs in this
+extension.  If the application fails to do this, the implementation throws
+a synchronous exception with the `errc::kernel_not_supported` error code
+when the kernel is submitted to the queue.
+
+_Occasionally, an extension is limited to certain backends and there is no
+related device aspect. When this is the case, include a paragraph like:_
+
+This extension is currently implemented in {dpcpp} only for GPU devices and
+only when using the Level Zero backend.  Attempting to use this extension in
+kernels that run on other devices or backends may result in undefined
+behavior.  Be aware that the compiler is not able to issue a diagnostic to
+warn you if this happens.
+
+== Overview
+
+Although SYCL supports task-level parallelism through device kernels, there is
+no specific mechanism to express parallelism at a sub-kernel level or to invoke
+additional kernels without involving the host program. In OpenCL 2.0 and beyond,
+kernel invocations can be initiated without host program interaction through
+device side enqueue. C++11 also introduced `std::async()` for task-level
+parallelism. However, neither of these approaches is a good fit for spatial
+compute platforms, which require a clear mapping of parallel constructs to
+spatial instances in the hardware.
+
+Rather than significantly changing the semantics of existing solutions, this
+extension introduces a sub-kernel task-level parallelism interface to SYCL for
+use in device code. It provides the class
+`sycl::ext::intel::experimental::task_sequence`, which is an
+abstraction of an asynchronous sequence of invocations of a callable that can be
+executed asynchronously from the caller, and with which a clear mapping to
+spatial hardware instances can be reasoned about when developing applications.
+
+NOTE: In this document, we use `task_sequence` to indicate the proposed
+`sycl::ext::intel::experimental::task_sequence`
+
+== Definitions
+
+== `task_sequence` class
+
+[source,c++,linenums]
+----
+namespace sycl {
+namespace ext {
+namespace intel {
+namespace experimental {
+
+template<auto& f, class propertiesT =
+   decltype(oneapi::experimental::properties{})>
+class task_sequence{};
+
+template <typename ReturnT, typename... ArgsT, ReturnT (&f)(ArgsT...),
+          class propertiesT>
+class task_sequence<f, propertiesT> {
+
+public:
+  task_sequence(const task_sequence &) = delete;
+  task_sequence &operator=(const task_sequence &) = delete;
+  task_sequence(task_sequence &&) = delete;
+  task_sequence &operator=(task_sequence &&) = delete;
+
+  task_sequence();
+
+  task_sequence(const propertiesT &);
+
+  void async(ArgsT... args);
```

Comment:
Do not use a pre-C++11 syntax. `std::thread` and `std::async` have this because it was designed before C++11 and the lambdas... I guess we do not want this technical debt in 2023.
What about a syntax taking only a lambda in your system?
This would avoid passing a function reference?
How to use your system with a lambda?
But at the end, give me a compelling example. :-)

Reply:
We did consider using lambdas. However we discarded it for reasons related to determining static resolvability of the function.